scrapy/spiders/crawl.py

```diff
@@ -19,17 +19,28 @@ def identity(x):
 
 class Rule(object):
 
-    def __init__(self, link_extractor, callback=None, cb_kwargs=None, follow=None, process_links=None, process_request=identity):
+    def __init__(self, link_extractor, callback=None, cb_kwargs=None, follow=None, process_links=None, process_request=None):
         self.link_extractor = link_extractor
         self.callback = callback
         self.cb_kwargs = cb_kwargs or {}
         self.process_links = process_links
-        self.process_request = process_request
+        self.process_request = process_request or identity
         if follow is None:
             self.follow = False if callback else True
         else:
             self.follow = follow
 
+    def _process_request(self, request, response):
+        """
+        Wrapper around the request processing function to maintain backward compatibility
+        with functions that do not take a Response object as parameter.
+        """
+        argcount = self.process_request.__code__.co_argcount
```

Comment:
Very elegant solution :)

Reply:
What about a deprecation warning when the specified function only expects one argument?

Reply:
That's a good idea, I'll get to it, thanks!

Reply:
There are also some related utilities in `scrapy.utils.python`: `get_func_args` and `get_spec`. I wonder if we can use an existing function. If not, I think this code should be extracted to a new helper in scrapy.utils.python, with dedicated tests for it.

Reply:
Awesome, `get_func_args` did the trick.